File: 20190519python/0463. Island Perimeter.py
class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col]==1:
                    return self.calIslandPerimeter(grid,row,col)
        
    def calIslandPerimeter(self,grid,row,col):
        rep=4
        grid[row][col]=2
        if row-1>=0:
            if grid[row-1][col]==1:
                rep+=self.calIslandPerimeter(grid,row-1,col)-1
            elif grid[row-1][col]==2:
                rep-=1
        if col-1>=0:
            if grid[row][col-1]==1:
                rep+=self.calIslandPerimeter(grid,row,col-1)-1
            elif grid[row][col-1]==2:
                rep-=1
        if row+1<len(grid):
            if grid[row+1][col]==1:
                rep+=self.calIslandPerimeter(grid,row+1,col)-1
            elif grid[row+1][col]==2:
                rep-=1
        if col+1<len(grid[0]):
            if grid[row][col+1]==1:
                rep+=self.calIslandPerimeter(grid,row,col+1)-1
            elif grid[row][col+1]==2:
                rep-=1
        return rep

File: archives/C#/0463. Island Perimeter.cs
public class Solution {
    public int IslandPerimeter(int[,] grid) {
        for(int row=0;row<grid.GetLength(0);row++){
            for(int column=0;column<grid.GetLength(1);column++){
                if(grid[row,column]==1){
                    return CalIslandPerimeter(grid,row,column);
                }
            }
        }
        return 0;
    }
    public int CalIslandPerimeter(int[,] grid,int row,int column){
        int rep=4;
        grid[row,column]='X';
        if (row-1>=0){
            if(grid[row-1,column]==1){
                rep+=CalIslandPerimeter(grid,row-1,column)-1;
            }
            else if (grid[row-1,column]=='X'){
                rep--;
            }
        }
        if (column-1>=0){
            if(grid[row,column-1]==1){
                rep+=CalIslandPerimeter(grid,row,column-1)-1;
            }
            else if (grid[row,column-1]=='X'){
                rep--;
            }
        }
        if (row+1<grid.GetLength(0)){
            if(grid[row+1,column]==1){
                rep+=CalIslandPerimeter(grid,row+1,column)-1;
            }
            else if (grid[row+1,column]=='X'){
                rep--;
            }
        } 
        if (column+1<grid.GetLength(1)){
            if(grid[row,column+1]==1){
                rep+=CalIslandPerimeter(grid,row,column+1)-1;
            }
            else if (grid[row,column+1]=='X'){
                rep--;
            }
        }
        return rep;
    }
}
